processor/src/main/java/org/mapstruct/ap/model/dependency/GraphAnalyzer.java

```diff
@@ -0,0 +1,134 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.model.dependency;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+/**
+ * Analyzes graphs: Discovers all descendants of given nodes and detects cyclic dependencies between nodes if present.
+ *
+ * @author Gunnar Morling
+ */
+public class GraphAnalyzer {
+
+    private final Map<String, Node> nodes = new HashMap<String, Node>();
+    private final Set<List<String>> cycles = new HashSet<List<String>>();
+
+    private final Stack<Node> currentPath = new Stack<Node>();
+
+    public void addNode(String name, List<String> descendants) {
+        Node node = getNode( name );
+
+        for ( String descendant : descendants ) {
+            node.addDescendant( getNode( descendant ) );
+        }
+    }
+
+    public void addNode(String name, String... descendants) {
+        Node node = getNode( name );
+
+        for ( String descendant : descendants ) {
+            node.addDescendant( getNode( descendant ) );
+        }
+    }
+
+    /**
+     * Performs a full traversal of the graph, detecting potential cycles and calculates the full list of descendants of
+     * the nodes.
+     */
+    public void analyze() {
+        for ( Node node : nodes.values() ) {
+            depthFirstSearch( node );
+        }
+    }
+
+    /**
+     * Returns all the descendants of the given node, either direct or transitive ones.
+     * <p>
+     * <b>Note</b>:The list will only be complete if the graph contains no cycles.
+     */
+    public Set<String> getAllDescendants(String name) {
+        Node node = nodes.get( name );
+        return node != null ? node.getAllDescendants() : Collections.<String>emptySet();
+    }
+
+    public Set<List<String>> getCycles() {
+        return cycles;
+    }
+
+    private void depthFirstSearch(Node node) {
+        if ( node.isProcessed() ) {
+            return;
+        }
+
+        currentPath.push( node );
+
+        // the node is on the stack already -> cycle
+        if ( node.isVisited() ) {
+            cycles.add( getCurrentCycle( node ) );
+            currentPath.pop();
+            return;
+        }
+
+        node.setVisited( true );
+
+        for ( Node descendant : node.getDescendants() ) {
+            depthFirstSearch( descendant );
+            node.getAllDescendants().addAll( descendant.getAllDescendants() );
+        }
+
+        node.setProcessed( true );
+        currentPath.pop();
+    }
+
+    private List<String> getCurrentCycle(Node start) {
+        List<String> cycle = new ArrayList<String>();
+        boolean inCycle = false;
+
+        for ( Node n : currentPath ) {
+            if ( n.getName().equals( start.getName() ) ) {
```

Comment:
you could even check for ```n == start```, right? Or use ```n.equals( start )```, as your equals-implementation in Node checks for reference equality first.